YAML and Python script for the AHI CSR BUFR

PraveenKumar-NOAA opened the pull request:
This PR introduces the following files for the AHI CSR BUFR converter:

- ~/dump/mapping/bufr_ahicsr.yaml
- ~/dump/mapping/bufr_ahicsr.py
- ~/ush/test/config/bufr_bufr4backend_ahicsr.yaml
- ~/ush/test/config/bufr_script4backend_ahicsr.yaml


**Testing and Validation**

- All four workflows (bufr2netcdf, bufr4backend, script2netcdf, and script4backend) were executed and tested using serial execution.

- Validation tests were completed for the following variables:

-- brightnessTemperature
-- cloudAmount

The validation was performed by comparing the input BUFR files and the corresponding output IODA files. Additional details can be found in #41.


**Python Coding Standards**

- The spoc_python_coding_norms test was run to check compliance with Pycodestyle standards.
- Result: All checks passed successfully (100% pass, 0 failures).

**Variable Naming Validation**

- The variable naming convention test is currently failing due to the absence of the following variables in the naming convention table. To resolve this, these variables need to be added to the table.

1. Variable 'MetaData/numberPixelVertical'
2. Variable 'MetaData/sensorBandWidth'
3. Variable 'MetaData/reportIndicator'
4. Variable 'MetaData/radianceComputationalMethod'
5. Variable 'MetaData/satelliteInstrumentData' 
6. Variable 'MetaData/numberPixelHorizontal' 
7. Variable 'MetaData/radianceType' 
8. Variable 'MetaData/receiptTimeSignificance' 
9. Variable 'ObsValue/clearSkyStandardDeviation'_

All these variables have been removed from the YAML file, except clearSkyStandardDeviation, which still needs to be added to the ObsSpace.yaml file.

- A discussion will be scheduled with users and experts to determine if these variables should be included.

rmclaren committed:
got it working with obs_builder. trying to find additional improvements

rmclaren committed:
enhanced logging

rmclaren committed:
more ObsBuilder enhancements

rmclaren committed:
reorganized a little

rmclaren committed:
Merge branch 'develop' into feature/obs_builder

rmclaren committed:
updated satwnd amv files to latest ObsBuilder spec

rmclaren committed:
made way to get map path

rmclaren committed:
updated and fixed up avhrr to actually run

rmclaren committed:
initial implementation of prepbufr

rmclaren committed:
cleaned up some file names

rmclaren committed:
more updates

rmclaren committed:
improved get ref time

rmclaren committed:
improved get ref time

rmclaren committed:
improved get ref time

rmclaren committed:
improved get ref time

rmclaren committed:
improved get ref time

rmclaren committed:
improved get ref time

rmclaren committed:
removed obsolete flags from add_main_functions

emilyhcliu committed:
Update IASI, ATMS and Add CrIS. and update all Python scripts to follow pycodestye (#30)

**This PR updates two things:**
- Update mapping file and Python script for IASI, CrIS , ATMS 

   Update the following to IASI:**
   1. Rename mtiasi to iasi
   2. Change logging to self.log.warning
   3. Add map_path function

   Add new data type - cris-fsr**

   Update the following for ATMS:**
   1. Change logging to self.log.warning
   2. Add map_path function

- Add Python coding norm check (pycodestyle) --- this is an initial
implementation so we
have a coding norm check for Python script. This will be updated to use
pytest in another PR
   To use it:
   - Load obsforge modules (~eliu/modules/env_obsforge.sh)
   - go to ./sorc/spoc/build
   - ctest -VV -R spoc_python_coding_norms

- Modify all Python scripts to follow pycodestyle (ctest to be added in
another PR soon)
   (No change in science)

emilyhcliu committed:
Merge branch 'develop' into feature/obs_builder

emilyhcliu committed:
Add ASCAT (satellite-derived winds) mapping file and Python script (#37)

This PR addes two files for ASCAT
- mapping YAML
- Python  script  --- pass Python coding norm check

Notes:
- This PR should be tested with [bufr-query PR
#73](https://github.com/NOAA-EMC/bufr-query/pull/73)
- A new PR will be opened to modify the standalone AMV converters
accordingly after this PR is merged.

ASCAT data processing has the following features:
- Convert wind speed and direction to u and v wind components
- Add obs type (290)

Documentation is in Sphinx style 

Test: script2netcdf validated
Validation plots

IODA windEastward 


![gdas_ObsValue_windEastward_satwnd_ascat_metop-b](https://github.com/user-attachments/assets/3d71ed27-1d94-4df6-a3e1-a230dca6a9d1)

GSI windEastward


![gdas_ObsValue_windEastward_ascatw_ascat_metop-b](https://github.com/user-attachments/assets/35473373-4236-4031-985b-0c118529315c)


IODA windNorthward


![gdas_ObsValue_windNorthward_satwnd_ascat_metop-b](https://github.com/user-attachments/assets/a753cdfe-2bac-461f-a418-078ac0ceaa41)


GSI windNorthward

![gdas_ObsValue_windNorthward_ascatw_ascat_metop-b](https://github.com/user-attachments/assets/5e49b097-007a-46d7-a9c2-0c13274fd819)

---------

Co-authored-by: Emily Liu <eliu@hercules-login-4.hpc.msstate.edu>
Co-authored-by: Emily Liu <eliu@hercules-login-3.hpc.msstate.edu>

emilyhcliu committed:
Add SSMIS mapping file and python scripts (#36)

This PR adds two files for SSMIS
- mapping YAML
- Python script   --- pass Python coding norm test

Notes:
- This PR should be tested with [bufr-query PR
#74](https://github.com/NOAA-EMC/bufr-query/pull/74)

SSMIS data processing has the following features:
- spatial averaging (this is implemented in bufr-query variable
transform) and it can be used by activated the feature from YAML
- Add new variables via the obs builder Python API
- satellite ascending/descending orbit flag (1 for ascending and -1 for
descending)
    - solar zenith and azimuth angles (use pysolar)

Notes:
- Documentation is in Sphnix style (following discussion in Issue #35 )
- SSMIS data conversion includes spatial averaging. Therefore the
processing can only run with one MPI task.
- The solar angles calculation is done using pysolar. The functions used
do not support vectors. Therefore, the calculation of solar angles are
implemented with multiprocessing (e.g., one MPI task and spawned with 12
CPUs)

   ```
             srun -n 1 --cpus_per_tasks=12 python bufr_ssmis.py
   ```
   or 
   ```
             export CPUS_PER_TASKS=12
             python bufr_ssmis.py``
   ```
   
Test: script2netcdf validated
Plots for output netCDF from SSMIS converter   
 

![gdas_solarZenithAngle_ssmis_f17_channel_4](https://github.com/user-attachments/assets/8a90a7c6-9f69-435a-8d15-46b3aa6e698d)


![gdas_satelliteAscendingFlag_ssmis_f17_channel_4](https://github.com/user-attachments/assets/c583697f-7ce7-435c-969b-1ef953fcd090)


![gdas_ObsValue_ssmis_f17_channel_4](https://github.com/user-attachments/assets/74e80396-4865-4ffa-a18f-3bf822ff98d9)

---------

Co-authored-by: Emily Liu <eliu@hercules-login-3.hpc.msstate.edu>
Co-authored-by: Emily Liu <eliu@hercules-login-4.hpc.msstate.edu>

Commit:
YAML and Python script for AHI-CSR

Commit:
Files for the bufr and script backends

Commit:
Removed a commented line from  bufr_script4backend_ahicsr.yaml file

xincjin-NOAA commented on dump/mapping/bufr_ahicsr.py:
```diff
@@ -0,0 +1,41 @@
+
+            satId = container.get('satelliteId', cat)
+            if not np.any(satId):
+                self.log.warning(f'Category {cat[0]} does not exist in input file')
```
Does this function do any real works?

PraveenKumar-NOAA replied:
@xincjin-NOAA thanks. Yes, it performs a validation check for each sub-category in the container.

xincjin-NOAA replied:
thanks

nicholasesposito commented on dump/mapping/bufr_ahicsr.yaml:
```diff
@@ -0,0 +1,247 @@
+
+    - name: "source"
+      type: string
+      value: "U.S. National Weather Service, National Centres for Environmental Prediction (NCEP)"
```
"Centers" not Centres

PraveenKumar-NOAA replied:
@nicholasesposito thanks, fixed.

nicholasesposito commented on dump/mapping/bufr_ahicsr.yaml:
```diff
@@ -0,0 +1,247 @@
+
+    - name: "providerFullName"
+      type: string
+      value: "Japan Meteorological Agencyi (JMA)"
```
"Agency"

PraveenKumar-NOAA replied:
@nicholasesposito thanks, fixed.

nicholasesposito commented on ush/test/config/bufr_bufr4backend_ahicsr.yaml:
```diff
@@ -0,0 +1,22 @@
+        category: ["h8"]
+        cache categories:            # optional
+          - ["h8"]
+          - ["h9"]
```
Am I correct in assuming we're not doing h9?

PraveenKumar-NOAA replied:
@nicholasesposito thanks for pointing this. We need both, h8 and h9. The backend YAMls were updated for this change.

Commit:
Updated backend YAMLs for category h9

Commit:
Fixes spelling in bufr_ahicsr.yaml

emilyhcliu committed:
Update YAML and Python script for SSMIS, ATMS, IASI, and CrIS (#43)

### This PR updates the YAML and Python scripts for SSMIS ATMS, IASI and
CrIS

For all four sensors:
1. Simplify the handling of missing categories using the
`add_dummy_variable` function
2. Use `map_path` from import.obs_builder for ATMS, CrIS, and IASI
3. Remove "type: netcdf" from the encoder section in the YAML
4. Updates `globals` in YAML
5. The output files were validated against the IODA validator with
ObsSpace.yaml; new variables were added to obsSpace.yaml. This validator
checks if the variables in the file follow IODA convention in terms of
name, unit, and dimension.

Here is an example of the output from running the ioda validator for
SSMIS output
```
Reading YAML from ../obsForge/sorc/ioda/share/ioda/yaml/validation/ObsSpace.yaml
Processing data file: ./testoutput/2024021900/script2netcdf/gdas.t00z.ssmis_f17.tm00.nc
 Verifying that all required groups exist
 Verifying group MetaData
 Verifying group ObsValue
 Verifying group /
 Verifying dimension names
 Verifying variable information
  Variable MetaData/earthSurfaceType
  Variable MetaData/sensorZenithAngle
  Variable MetaData/satelliteIdentifier
  Variable MetaData/sensorScanPosition
  Variable MetaData/longitude
  Variable MetaData/latitude
  Variable MetaData/sensorViewAngle
  Variable MetaData/scanLineNumber
  Variable MetaData/solarZenithAngle
  Variable MetaData/sensorChannelNumber
  Variable MetaData/satelliteAscendingFlag
  Variable MetaData/sensorAzimuthAngle
  Variable MetaData/rainFlag
  Variable MetaData/dateTime
  Variable MetaData/solarAzimuthAngle
  Variable ObsValue/brightnessTemperature
Final results:
  # errors:      0
  # warnings:    0

```

Notes:
- SSMIS specific - ssmis conversion uses updated `comupte_solar_angles`
which is under review in [bufr-query PR
#96](https://github.com/NOAA-EMC/bufr-query/pull/96)
- The update of ObsSpace.yaml from IODA is ongoing. We will add new
variables in the [ObsSpace.yaml
](https://github.com/JCSDA-internal/ioda/blob/feature/ObsSpace_Validator/share/ioda/yaml/validation/ObsSpace.yaml)in
IODA branch feature/ObsSpace_Validator

Specific to IASI and CrIS
Modifications made to cloud related variables - `cloudHeight`,
`cloudCoverTotal`, `fractionOfClearInFOV`:
- change group from `MetaData` to `ObsValue`
- change variable type from `int` to `float`
- scale the variables so they range between 0 and 1
- change `ObsValue/radiance` to `ObsValue/spectralRadiance` and unit
accordingly

The above changes are to following IODA convention.



These changes do not alter the results (scientific)

---------

Co-authored-by: Emily Liu <eliu@hercules-login-3.hpc.msstate.edu>
Co-authored-by: Emily Liu <eliu@hercules-login-4.hpc.msstate.edu>

emilyhcliu committed:
Merge branch 'develop' into feature/obs_builder

emilyhcliu committed:
Merge branch 'feature/obs_builder' into feature/obs_builder_ahicsr

emilyhcliu commented:
@PraveenKumar-NOAA 
I tested the` ahicsr` converter.
The channel dimension is missing in the IODA file (output)
```
netcdf gdas.t00z.ahi_h9.tm00 {
dimensions:
        Location = 317847 ;
variables:
        int Location(Location) ;
                Location:_FillValue = 2147483647 ;

```

This is because your YAML file does not include the dimension section in the encoder section.

Here is the example from ATMS:
https://github.com/NOAA-EMC/spoc/blob/369e647dcad24067f365b5c75b83bc74115abb58/dump/mapping/bufr_atms.yaml#L75-L80

emilyhcliu commented on dump/mapping/bufr_ahicsr.yaml:
```diff
@@ -0,0 +1,247 @@
+          _173: h8 
+          _174: h9
+
+encoder:
+
```
You need to include `dimension` section 

Here is the example from ATMS:
https://github.com/NOAA-EMC/spoc/blob/369e647dcad24067f365b5c75b83bc74115abb58/dump/mapping/bufr_atms.yaml#L75-L80

PraveenKumar-NOAA replied:
@emilyhcliu thanks, done.

Commit:
Delete bufr_adpupa_prepbufr.py and bufr_adpupa_prepbufr_mapping.yaml

Commit:
Delete bufr_adpupa_prepbufr.py and bufr_adpupa_prepbufr_mapping.yaml

Commit:
Updated YAML and Python scripts for the Channel dimension and sensorChannelNumber variable

PraveenKumar-NOAA commented:
> @PraveenKumar-NOAA I tested the` ahicsr` converter. The channel dimension is missing in the IODA file (output)
> 
> ```
> netcdf gdas.t00z.ahi_h9.tm00 {
> dimensions:
>         Location = 317847 ;
> variables:
>         int Location(Location) ;
>                 Location:_FillValue = 2147483647 ;
> ```
> 
> This is because your YAML file does not include the dimension section in the encoder section.
> 
> Here is the example from ATMS:
> 
> https://github.com/NOAA-EMC/spoc/blob/369e647dcad24067f365b5c75b83bc74115abb58/dump/mapping/bufr_atms.yaml#L75-L80

@emilyhcliu thanks. A dimension section is added in the encoder section for the Channel dimension, there are 12 Channels.

Commit:
Updated Channel Numbers from 7 to 16 and removed some extra variables from YAML

Commit:
Removed white spaces from the Python script

PraveenKumar-NOAA commented:
@emilyhcliu A brightness temperature plot has been created for the HIMAWARI-8 CLEAR SKY RADIANCES channel and added to the following issue: https://github.com/NOAA-EMC/spoc/issues/41. This completes the validation for this converter, as the test input BUFR file contains only HIMAWARI-8 data.

HyundeokChoi-NOAA commented:
This PR was merged to PR https://github.com/NOAA-EMC/spoc/pull/101
